### PR TITLE
Add XIAO_ESP32C3 and connect by built-in USB

### PR DESCRIPTION
**Pull request by matsujirushi**
## Description

I want to run nanoFramework with the built-in USB of ESP32C3.

## Motivation and Context

- Closes nanoFramework/Home#1225

## How Has This Been Tested?

Tested with Seeed Studio XIAO ESP32C3 and Visual Studio 2022.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

File: targets/ESP32/_common/WireProtocol_HAL_Interface.c [hunk 1]
```diff
@@ -161,21 +320,8 @@ void WP_ReceiveBytes(uint8_t **ptr, uint32_t *size)
     // check for request with 0 size
     if (*size)
     {
-
-#if CONFIG_USB_CDC_ENABLED
```
[the remark this answers is not in the document]
Comment (matsujirushi) [reply targets/ESP32/_common/WireProtocol_HAL_Interface.c:165]: @josesimoes 
Thanks for review.
Is it OK to add `&& defined(CONFIG_IDF_TARGET_ESP32S2)` here?
https://github.com/matsujirushi/nf-interpreter/blob/84962a64383f7692561a8cee152250ba5b5a4f9d/targets/ESP32/_common/WireProtocol_HAL_Interface.c#L174

File: targets/ESP32/_nanoCLR/targetHAL_Power.c [hunk 2]
```diff
@@ -5,10 +5,14 @@
 
 #include <esp32_idf.h>
 #include <nanoHAL_v2.h>
+#include <soc/rtc_cntl_reg.h>
 
 inline void CPU_Reset()
 {
-    esp_restart();
+    SET_PERI_REG_MASK(RTC_CNTL_OPTIONS0_REG, RTC_CNTL_SW_SYS_RST);
```
[the remark this answers is not in the document]
Comment (matsujirushi) [reply targets/ESP32/_nanoCLR/targetHAL_Power.c:12]: After stopping debugging in Visual Studio, USB-CDC became unable to communicate. ( espressif/esp-idf#10979 )
Therefore I changed to Software system reset (Core Reset level). See P182 in [ESP32-C3 TRM](https://www.espressif.com/sites/default/files/documentation/esp32-c3_technical_reference_manual_en.pdf).

File: targets/ESP32/CMakePresets.json [hunk 3]
```diff
@@ -258,6 +258,29 @@
                 "ESP32_RESERVED_RAM_FOR_IDF_ALLOCATION": "100"
             }
         },
+        {
+            "name": "XIAO_ESP32C3",
```
[the remark this answers is not in the document]
Comment (matsujirushi) [reply targets/ESP32/CMakePresets.json:262]: Added.

File: targets/ESP32/_common/WireProtocol_HAL_Interface.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (matsujirushi) [reply targets/ESP32/_common/WireProtocol_HAL_Interface.c:165]: I've added `&& CONFIG_IDF_TARGET_ESP32S2` but haven't been able to test it.

File: targets/ESP32/_nanoCLR/targetHAL_Power.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (josesimoes) [reply targets/ESP32/_nanoCLR/targetHAL_Power.c:12]: The correct way is to use the IDF API. There are other tasks that are performed besides "just" reseting the core.
Let's wait for the feedabck from the IDF team on this.

In the meantime, let's have them both with a couple of if defs to checl for the USB-CDC and the other variants.

**Comment by josesimoes** [comment]
@matsujirushi any chance you can address the latest review comments/requests to have this merged?
It's soooo close!

File: targets/ESP32/_nanoCLR/targetHAL_Power.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (matsujirushi) [reply targets/ESP32/_nanoCLR/targetHAL_Power.c:12]: I'm sorry for being late.

> let's have them both with a couple of if defs to checl for the USB-CDC and the other variants.

I didn't understand it properly.
Will update soon.

[the remark this answers is not in the document]
Comment (matsujirushi) [reply targets/ESP32/_nanoCLR/targetHAL_Power.c:14]: Updated.

**Comment by matsujirushi** [comment]
@josesimoes How is your progress?

**Review by josesimoes — APPROVED** [review]
LGTM!
I think we are in good shape to release this. 
@matsujirushi thanks for taking care of it. Trust you had fun and learned something along the way. 😉

**Pull request merged**